### PR TITLE
DNS seeder integration.

### PR DESCRIPTION
**Pull request by fchirica**

### CODE REVIEW

File: chia/server/node_discovery.py [hunk 1]
```diff
@@ -166,12 +169,29 @@ async def on_connect(peer: ws.WSChiaConnection):
 
         await self.server.start_client(self.introducer_info, on_connect)
 
+    async def _query_dns(self, dns_address):
+        try:
+            peers: List[TimestampedPeerInfo] = []
+            result = dns.resolver.query(dns_address, "A")
```
Comment (arvidn) [inline chia/server/node_discovery.py:175]: isn't this async?
DNS lookups can hang for 2 minutes in the worst case. I think it's important to not make a blocking call.

Also, shouldn't we look up AAAA records as well?
Even though the server doesn't support it yet, I expect we'll add IPv6 support at some point

File: chia/server/node_discovery.py [hunk 2]
```diff
@@ -182,22 +202,29 @@ async def _connect_to_peers(self, random) -> None:
...
-                    # that once we get peers, we stop contacting the introducer.
-                    try:
-                        await asyncio.sleep(5)
-                    except asyncio.CancelledError:
-                        return None
-                    empty_tables = False
+                    # Run dual between DNS servers and introducers. One time query DNS server,
+                    # next two times query the introducer.
+                    if introducer_attempts % 3 == 0 and len(self.dns_servers) > 0:
+                        dns_address = random.choice(self.dns_servers)
```
Comment (arvidn) [inline chia/server/node_discovery.py:213]: it seems round-robin would be better here. if one DNS server is down or broken, we want to test the next one, never the same one again. We could still shuffle the DNS server list to make the order we try them in random.

I think a simple way of implementing round-robin would be to rotate the list for every call

File: chia/server/node_discovery.py [hunk 3]
```diff
@@ -251,14 +278,17 @@ async def _connect_to_peers(self, random) -> None:
                 while not got_peer and not self.is_closed:
                     sleep_interval = 1 + len(groups) * 0.5
                     sleep_interval = min(sleep_interval, self.peer_connect_interval)
+                    # Special case: try to find our first peer much quicker.
+                    if len(groups) == 0:
+                        sleep_interval = 0.1
```
Comment (arvidn) [inline chia/server/node_discovery.py:283]: we won't try to connect the same peer more than once every 2 hours, is that correct? I think `sleep_interval` could be 0.1 seconds unconditionally, but I also think that's an unrelated change and maybe best reviewed in a separate PR.

(it would simplify some of this logic, because we could just make the default configuration for `peer_connect_interval` be 0.1 and just always use that)

File: chia/server/node_discovery.py [hunk 4]
```diff
@@ -284,7 +314,9 @@ async def _connect_to_peers(self, random) -> None:
                         addr = None
                         continue
                     # only consider very recently tried nodes after 30 failed attempts
-                    if now - info.last_try < 600 and tries < 30:
+                    # attempt a node once per 2 hours if we lack connections to increase the chance
+                    # to try all the peer table.
+                    if now - info.last_try < 2 * 3600 and tries < 30:
```
Comment (arvidn) [inline chia/server/node_discovery.py:319]: somewhat unrelated: is `now` posix time as UTC? It's important it doesn't have daylight savings for instance, we need it to be monotonic.

File: chia/util/initial-config.yaml [hunk 5]
```diff
@@ -287,6 +287,10 @@ full_node:
   # timeout for weight proof request
   weight_proof_timeout: 360
 
+  # List of trusted DNS seeders to bootstrap from.
+  # If you modify this, please change the hardcode as well from FullNode.set_server()
+  dns_servers:
+    - "seeder.fchirica.com"
```
Comment (arvidn) [inline chia/util/initial-config.yaml:293]: we should update this to `seeder.chia.net` before we land this, I think.

**Review by arvidn — CHANGES_REQUESTED** [review]
it looks like dnspython does support asynchronous queries.

https://dnspython.readthedocs.io/en/stable/async-resolver-class.html

File: chia/util/initial-config.yaml (hunk 5 above)
Comment (hoffmang9) [reply chia/util/initial-config.yaml:293]: We actually want a list of approved dns servers that will include a chia.net one and eg a Hoffmang.com one etc. I have a couple of volunteers lined up

**Comment by hoffmang9** [comment]
And we need to support IPv6 initially.

File: chia/server/node_discovery.py (hunk 3 above)
Comment (fchirica) [reply chia/server/node_discovery.py:283]: we need to at least profile that and see if it doesn't interfere with other cpu-intensive functions (for example sync). on the other side, if you have 0 peers, it's safe to keep the sleep small, as there's not too much else going on.

**Comment by arvidn** [comment]
this looks good to me. but I really think you should drop the last commit where you merge `main` into this branch.

**Review by arvidn — APPROVED** [review]
